### DISCUSSION TITLE
How do I parse OffsetDateTime from bytes instead of str?

### DISCUSSION THREAD

**Discussion by Nugine** [time-rs/time#540] [Question]
```rust
fn parse_rfc3339(input: &[u8]) -> Result<OffsetDateTime, time::error::Parse> {
    todo!()
}
```

**Comment by jhpratt** [comment]
RFC3339 necessarily requires valid ASCII (and thus UTF-8). Why do you need to parse from raw bytes? It's possible, but it's more complicated _because_ it isn't normally what you want.

**Reply by Nugine** [reply]
S3 protocol passes some timestamp fields via http headers. I don't have to perform unnecessary utf-8 validation if I can parse them from bytes.
See also <https://docs.rs/hyper/0.14.23/hyper/header/struct.HeaderValue.html#method.to_str>.

(This optimization is not critical, but why not?)

**Reply by jhpratt** [reply]
Sure. In my opinion it's premature optimization, but to each their own.

In short, you will have to go through the `Parsed` struct. There are a number of methods that will let you parse the raw format description items. Note that you will have to handle any trailing characters yourself. Once the `Parsed` struct is complete, you can use the `TryFrom` implementation to obtain the final value.

**Reply by Nugine** [reply]
It seems too complicated for me. I would turn to ascii validation.

**Reply by jhpratt** [reply]
As I said, it's more complicated because it's normally not what's desirable. Ultimately it's the exact thing that `time` does for strings, just abstracted away.
